acceptance-tests/src/test/java/tech/pegasys/ethsigner/tests/tls/ClientSideTlsAcceptanceTest.java

```diff
@@ -0,0 +1,196 @@
+      throws IOException {
+
+    final Path clientPasswordFile =
+        Files.writeString(workDir.resolve("clientKeystorePassword"), presentedCert.getPassword());
+    final Path serverPasswordFile =
+        Files.writeString(
+            workDir.resolve("clientTrustStorePassword"), expectedWeb3ProviderCert.getPassword());
+
+    final SignerConfigurationBuilder builder = new SignerConfigurationBuilder();
+    builder.withDownstreamTrustStore(
+        new BasicPkcsStoreConfig(
+            expectedWeb3ProviderCert.getPkcs12File(), serverPasswordFile.toFile()));
+    builder.withDownstreamKeyStore(
+        new BasicPkcsStoreConfig(presentedCert.getPkcs12File(), clientPasswordFile.toFile()));
+    builder.withHttpRpcPort(listenPort);
+
+    final NodeConfiguration nodeConfig = new NodeConfigurationBuilder().build();
+    final NodePorts nodePorts = new NodePorts(downstreamWeb3Port, 9);
+
+    final Signer signer = new Signer(builder.build(), nodeConfig, nodePorts);
+
+    return signer;
+  }
+
+  @Test
+  void ethSignerProvidesSpecifiedClientCertificateToDownStreamServer(@TempDir Path workDir)
+      throws IOException {
+
+    final TlsCertificateDefinition serverCert =
+        TlsCertificateDefinition.loadFromResource("tls/cert1.pfx", "password");
+    final TlsCertificateDefinition ethSignerCert =
+        TlsCertificateDefinition.loadFromResource("tls/cert2.pfx", "password2");
+
+    // Note: the HttpServer always responds with a JsonRpcSuccess, result=300.
+    final HttpServer web3ProviderHttpServer =
+        serverFactory.create(serverCert, ethSignerCert, workDir);
+
+    final Signer signer =
+        createAndStartSigner(
+            ethSignerCert, serverCert, web3ProviderHttpServer.actualPort(), 0, workDir);
+
+    assertThat(signer.accounts().balance("0x123456"))
+        .isEqualTo(BigInteger.valueOf(MockBalanceReporter.REPORTED_BALANCE));
+  }
+
+  @Test
+  void ethSignerDoesNotConnectToServerNotSpecifiedInTrustStore(@TempDir Path workDir)
+      throws IOException {
+    final TlsCertificateDefinition serverPresentedCert =
+        TlsCertificateDefinition.loadFromResource("tls/cert1.pfx", "password");
+    final TlsCertificateDefinition ethSignerCert =
+        TlsCertificateDefinition.loadFromResource("tls/cert2.pfx", "password2");
+    final TlsCertificateDefinition ethSignerExpectedServerCert =
+        TlsCertificateDefinition.loadFromResource("tls/cert2.pfx", "password2");
+
+    final HttpServer web3ProviderHttpServer =
+        serverFactory.create(serverPresentedCert, ethSignerCert, workDir);
+
+    final Signer signer =
+        createAndStartSigner(
+            ethSignerCert,
+            ethSignerExpectedServerCert,
+            web3ProviderHttpServer.actualPort(),
+            0,
+            workDir);
+
+    assertThatThrownBy(() -> signer.accounts().balance("0x123456"))
+        .isInstanceOf(ClientConnectionException.class)
+        .hasMessageContaining(String.format("%d", BAD_GATEWAY.code()));
```

Comment:
String.valueOf(BAD_GATEWAY.code()) might have been a bit faster than the format :)

Reply:
done